### PR TITLE
Refactor SDKv2 and PF HCL utilities to expose a shared interface

### PR DESCRIPTION
**Pull request by VenelinMartinov**
This change refactors the HCL writing utilities to expose a shared interface on which we can implement the writing capabilities.

I've opted to make a new HCL shim layer because:
 - We shouldn't complicate the runtime behaviour with test code
 - The HCL writing is simpler than the runtime behaviour so we can get away with a much simpler interface which only cares about Attributes, Blocks and nesting
 

We did discuss that this is not necessary in https://github.com/pulumi/pulumi-terraform-bridge/pull/2577 but I then hit the need to implement TF lifecycle methods for PF which is already done for SDKv2, so this makes it possible to implement such shared concerns once.

### CODE REVIEW

**Comment by VenelinMartinov** [comment]
This change is part of the following stack:

- #2589 ◀

<sub>Change managed by [git-spice](https://abhinav.github.io/git-spice/).</sub>

File: pkg/internal/tests/cross-tests/impl/hclwrite.go [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+		default:
+			contract.Failf("unexpected nesting mode %v", block.GetNestingMode())
+		}
+	}
+
+	// lifecycle block
+	if _, ok := config["lifecycle"]; ok {
+		newBlock := body.AppendNewBlock("lifecycle", nil)
+		writeBlock(newBlock.Body(), &lifecycleBlock{}, config["lifecycle"].AsValueMap())
+	}
```
Comment (iwahbe) [inline pkg/internal/tests/cross-tests/impl/hclwrite.go:159]: I'm assuming that we only want `"lifecycle"` to be special at the top level. This will treat it specially at all levels. Given that, can we move this speciality out of `writeBlock` and into `writeResource`. `lifecycle` can just be an unconditional addition to the set of blocks at the top level.

File: pkg/internal/tests/cross-tests/tfwrite.go [hunk 2]
```diff
@@ -16,115 +16,70 @@
...
-	_, err := f.WriteTo(out)
-	return err
+type hclSchemaSDKv2 struct {
+	sch map[string]*schema.Schema
 }
 
-func writeBlock(body *hclwrite.Body, schemas map[string]*schema.Schema, values map[string]cty.Value) {
-	internalMap := schema.InternalMap(schemas)
-	coreConfigSchema := internalMap.CoreConfigSchema()
+func NewHCLSchemaSDKv2(sch map[string]*schema.Schema) crosstestsimpl.ShimHCLSchema {
```
Comment (iwahbe) [inline pkg/internal/tests/cross-tests/tfwrite.go:41]: I don't think we need a helper function here, and it certainly shouldn't be public.

File: pkg/internal/tests/cross-tests/tfwrite.go [hunk 3]
```diff
@@ -16,115 +16,70 @@
...
-		return fmt.Errorf("WriteHCL cannot yet write unknowns")
-	}
-	f := hclwrite.NewEmptyFile()
-	block := f.Body().AppendNewBlock("resource", []string{resourceType, resourceName})
-	writeBlock(block.Body(), sch, config.AsValueMap())
-	_, err := f.WriteTo(out)
-	return err
+type hclSchemaSDKv2 struct {
+	sch map[string]*schema.Schema
 }
```
Comment (iwahbe) [inline pkg/internal/tests/cross-tests/tfwrite.go:39]: Style nit: this can just be a type wrapper. It doesn't need to be an enclosing struct.
```suggestion
type hclSchemaSDKv2 map[string]*schema.Schema
```

File: pkg/internal/tests/cross-tests/tfwrite.go [hunk 4]
```diff
@@ -16,115 +16,70 @@
...
-			}
-		case "NestingMap":
-			contract.Failf("unexpected NestingMap for %s with schema %s", key, schemas[key].GoString())
-		default:
-			contract.Failf("unexpected nesting mode %v", bl.Nesting)
-		}
+func (s hclSchemaSDKv2) Attributes() map[string]crosstestsimpl.ShimHCLAttribute {
+	internalMap := schema.InternalMap(s.sch)
+	coreConfigSchema := internalMap.CoreConfigSchema()
+	attrMap := make(map[string]crosstestsimpl.ShimHCLAttribute)
```
Comment (iwahbe) [inline pkg/internal/tests/cross-tests/tfwrite.go:50]:
```suggestion
	attrMap := make(map[string]crosstestsimpl.ShimHCLAttribute, len(s.sch))
```

File: pkg/internal/tests/cross-tests/tfwrite.go [hunk 5]
```diff
@@ -16,115 +16,70 @@
...
+	blockMap := make(map[string]crosstestsimpl.ShimHCLBlock)
+	for key := range coreConfigSchema.BlockTypes {
+		res := s.sch[key].Elem.(*schema.Resource)
+		nesting := coreConfigSchema.BlockTypes[key].Nesting.String()
+		blockMap[key] = hclBlockSDKv2{
+			hclSchemaSDKv2: hclSchemaSDKv2{
+				sch: res.Schema,
 			},
+			nesting: sdkV2NestingToShim(nesting),
 		}
```
Comment (iwahbe) [inline pkg/internal/tests/cross-tests/tfwrite.go:69]: We shouldn't need to re-index into the map whose keys we are already iterating over.
```suggestion
	for key, block := range coreConfigSchema.BlockTypes {
		res := s.sch[key].Elem.(*schema.Resource)
		blockMap[key] = hclBlockSDKv2{
			hclSchemaSDKv2: hclSchemaSDKv2{
				sch: res.Schema,
			},
			nesting: sdkV2NestingToShim(block.Nesting),
		}
```

Separate question: Why can't we use the schema from `block` instead of `s.sch[key].Elem.(*schema.Resource)`?

File: pkg/internal/tests/cross-tests/tfwrite.go [hunk 6]
```diff
@@ -16,115 +16,70 @@
...
+	crosstestsimpl "github.com/pulumi/pulumi-terraform-bridge/v3/pkg/internal/tests/cross-tests/impl"
 )
 
-// Provider writes a provider declaration to an HCL file.
-//
-// Note that unknowns are not yet supported in cty.Value, it will error out if found.
-func writeProvider(out io.Writer, sch map[string]*schema.Schema, typ string, config cty.Value) error {
-	if !config.IsWhollyKnown() {
-		return fmt.Errorf("WriteHCL cannot yet write unknowns")
+func sdkV2NestingToShim(nesting string) crosstestsimpl.Nesting {
```
Comment (iwahbe) [inline pkg/internal/tests/cross-tests/tfwrite.go:24]: This should take and switch on a schema.Nesting, not it's string representation.
```suggestion
func sdkV2NestingToShim(nesting schema.NestingMode) crosstestsimpl.Nesting {
```

File: pkg/pf/tests/internal/cross-tests/configure.go [hunk 7]
```diff
@@ -22,7 +22,7 @@ import (
 	"testing"
 
 	"github.com/hashicorp/terraform-plugin-framework/provider"
-	"github.com/hashicorp/terraform-plugin-framework/provider/schema"
+	pschema "github.com/hashicorp/terraform-plugin-framework/provider/schema"
```
Comment (iwahbe) [inline pkg/pf/tests/internal/cross-tests/configure.go:25]: Let's not do this. `pschema` often means Pulumi Schema, and we should package name overrides when possible unless they are standardized.

File: pkg/pf/tests/internal/cross-tests/configure.go [hunk 8]
```diff
@@ -109,7 +109,8 @@ func Configure(t T, schema schema.Schema, tfConfig map[string]cty.Value, options
 	// Run the TF part
 	{
 		var hcl bytes.Buffer
-		err := writeProvider(&hcl, schema, providerName, tfConfig)
+		sch := NewHCLSchemaPFProvider(schema)
+		err := crosstestsimpl.WriteProvider(&hcl, sch, providerName, tfConfig)
```
Comment (iwahbe) [inline pkg/pf/tests/internal/cross-tests/configure.go:113]: I don't see a reason to introduce intermediary state here.
```suggestion
		err := crosstestsimpl.WriteProvider(&hcl, NewHCLSchemaPFProvider(schema)
, providerName, tfConfig)
```

File: pkg/pf/tests/internal/cross-tests/tfwrite.go [hunk 9]
```diff
@@ -1,173 +1,182 @@
...
-) error {
-	if !cty.ObjectVal(config).IsWhollyKnown() {
-		return fmt.Errorf("WriteHCL cannot yet write unknowns")
+type pfNestingMode uint8
+
+const (
+	pfNestingModeUnknown pfNestingMode = 0
+	pfNestingModeList    pfNestingMode = 1
+	pfNestingModeSet     pfNestingMode = 2
+	pfNestingModeSingle  pfNestingMode = 3
```
Comment (iwahbe) [inline pkg/pf/tests/internal/cross-tests/tfwrite.go:16]: Unless the numbers matter (which they don't), then lets use `iota`.
```suggestion
	pfNestingModeUnknown pfNestingMode = iota
	pfNestingModeList    pfNestingMode 
	pfNestingModeSet     pfNestingMode 
	pfNestingModeSingle  pfNestingMode 
```

File: pkg/pf/tests/internal/cross-tests/tfwrite.go [hunk 10]
```diff
@@ -1,173 +1,182 @@
...
-
-		writePfObjectProvider(body, pschema.NestedBlockObject{
-			Attributes: schemas.Attributes,
-			Blocks:     schemas.Blocks,
-		}, value.AsValueMap())
 	default:
-		contract.Failf("Unknown block type: %T", schemas)
+		contract.Failf("Unknown block type: %T", block)
 	}
+	return pschema.NestedBlockObject{}
```
Comment (iwahbe) [inline pkg/pf/tests/internal/cross-tests/tfwrite.go:46]: Nit: lets move the `return` into the default block.
```suggestion
		contract.Failf("Unknown block type: %T", block)
		return pschema.NestedBlockObject{}
	}
```

File: pkg/pf/tests/internal/cross-tests/tfwrite.go [hunk 11]
```diff
@@ -1,173 +1,182 @@
...
+	return rschema.NestedBlockObject{}
+}
+
+type hclSchemaPFProvider struct {
+	sch pschema.Schema
+}
+
+var _ crosstestsimpl.ShimHCLSchema = hclSchemaPFProvider{}
+
+func NewHCLSchemaPFProvider(sch pschema.Schema) crosstestsimpl.ShimHCLSchema {
```
Comment (iwahbe) [inline pkg/pf/tests/internal/cross-tests/tfwrite.go:72]: I don't believe this wrapper is necessary, but it should definitely be private.
```suggestion
func newHCLSchemaPFProvider(sch pschema.Schema) crosstestsimpl.ShimHCLSchema {
```

**Review by iwahbe — COMMENTED** [review]
The change itself looks good. I left some notes on the implementation.

File: pkg/internal/tests/cross-tests/tfwrite.go (hunk 6 above)
Comment (VenelinMartinov) [reply pkg/internal/tests/cross-tests/tfwrite.go:24]: The schema.NestingMode is a private type, so we can't use it as a parameter here

File: pkg/internal/tests/cross-tests/tfwrite.go (hunk 5 above)
Comment (VenelinMartinov) [reply pkg/internal/tests/cross-tests/tfwrite.go:69]: Block doesn't contain the actual schema.Schema - it **should** contain the information we need though but we'd have to mess with the types and some of them are private. I think this is easier.

File: pkg/pf/tests/internal/cross-tests/tfwrite.go (hunk 9 above)
Comment (VenelinMartinov) [reply pkg/pf/tests/internal/cross-tests/tfwrite.go:16]: We do match the TF type here. I've added a comment to clarify that.

File: pkg/internal/tests/cross-tests/tfwrite.go (hunk 6 above)
Comment (VenelinMartinov) [reply pkg/internal/tests/cross-tests/tfwrite.go:24]: I'll use the same method I used in PF where we copied the type over.

File: pkg/pf/tests/internal/cross-tests/configure.go (hunk 7 above)
Comment (VenelinMartinov) [reply pkg/pf/tests/internal/cross-tests/configure.go:25]: pschema is used for provider schema throughout the PF code - I've renamed this to match the rest of the codebase.

The clash is unfortunate but I think we should address that seprately - in the PF code pschema means provider schema

File: pkg/internal/tests/cross-tests/tfwrite.go [hunk 12]
```diff
@@ -16,115 +16,75 @@
...
+	sdkV2NestingModeMap
+)
+
+func sdkV2NestingToShim(nesting sdkV2NestingMode) hclwrite.Nesting {
+	switch nesting {
+	case sdkV2NestingModeSingle:
+		return hclwrite.NestingSingle
+	case sdkV2NestingModeList:
+		return hclwrite.NestingList
+	case sdkV2NestingModeSet:
```
Comment (t0yv0) [inline pkg/internal/tests/cross-tests/tfwrite.go:43]: You're not handling sdkV2NestingModeMap with any warnings or such, worth a comment?

**Review by t0yv0 — APPROVED** [review]
I'd like to put forward an observation that instead of using tfshim.Schema to support tfwrite capability for SDKv1, SDKv2 and PF, this PR introduces a TF-write specific abstraction and a set of adapters to support SDKv2 and PF (not SDKv1). I have a suspicion that both are incomplete and have edge cases we are missing; we'll need to fix those twice. But what recommends this approach is that:

- no prod code is touched so we can move fast
- tests are fully independent of prod code which makes subsequent development much easier as we can go step by step, either doubting tests or doubting prod code, not both

I think overall this is the right call. Let's ship.

File: pkg/internal/tests/cross-tests/tfwrite.go (hunk 12 above)
Comment (VenelinMartinov) [reply pkg/internal/tests/cross-tests/tfwrite.go:44]: I believe the other nesting modes are not used. I have added an assert there with a message.
